### PR TITLE
[Merged by Bors] - feat(Tactic/Linter): the ppRoundtrip linter

### PR DESCRIPTION
**Pull request by adomani**
This linter tries to make sure that the source code looks like the pretty printed version of itself.

It is not intended to be always active, since not every pretty-printed syntax is desirable or type-correct!

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Tactic/Linter/PedanticLinter.lean
Comment (joneugster) [inline Mathlib/Tactic/Linter/PedanticLinter.lean:30]: Didn't you say this should be `false` by default? I assume this is for testing and will be changed before merging.

File: test/simp_confluence.lean
Comment (joneugster) [inline test/simp_confluence.lean:11]: If the linter is turned off by default, this isn't necessary

File: test/PedanticLinter.lean
Comment (joneugster) [inline test/PedanticLinter.lean:4]:
```suggestion
```

File: Mathlib/Tactic/Linter/PedanticLinter.lean
Comment (joneugster) [inline Mathlib/Tactic/Linter/PedanticLinter.lean:120]: Same as in another PR: imo it's barely worth when you can just write `Linter.getLinterValue linter.pedantic (← getOptions)` below...

File: Mathlib/Tactic/Linter/PedanticLinter.lean
Comment (joneugster) [inline Mathlib/Tactic/Linter/PedanticLinter.lean:66]: You can probably drop the debug comments now.

File: Mathlib/Tactic/Linter/PedanticLinter.lean
Comment (joneugster) [inline Mathlib/Tactic/Linter/PedanticLinter.lean:67]:
```suggestion
  let s := (split.map .trimLeft).filter (· != "")
```

File: Mathlib/Tactic/Linter/PedanticLinter.lean
Comment (joneugster) [inline Mathlib/Tactic/Linter/PedanticLinter.lean:128]:
```suggestion
    let stx := capSyntax stx (stx.getTailPos?.getD default).1
```
can the pedantic linter pedantic-lint itself? ;)

File: Mathlib/Tactic/Linter/PedanticLinter.lean
Comment (joneugster) [inline Mathlib/Tactic/Linter/PedanticLinter.lean:54]: ?

**Review by joneugster — APPROVED** [review]
I think this looks good, but as you said it should be disabled unless manually enabled.

File: Mathlib/Tactic/Linter/PedanticLinter.lean
Comment (joneugster) [reply Mathlib/Tactic/Linter/PedanticLinter.lean:120]: Update: I think to remember that `getLinterHash` was a problem in other PRs and you should directly call `Linter.getLinterValue linter.pedantic (← getOptions)`, do I remember that correctly?

**Comment by joneugster** [comment]
@adomani: I've pushed my suggestions myself now, adopting some responsibility for this PR, as I think it's a nice tool!

Therefore, somebody else will have to review it. (I might post in the correct Zulip channel about it)

**Comment by joneugster** [comment]
maintainer merge

(please have a bit a deeper look again: the last three commits by myself are essentially unreviewed -- mostly simple fixes)

**Comment by jcommelin** [comment]
Pedantic comment: what do you think of renaming this to something like `ppRoundTrip` instead of `pedantic`? I think that's a bit more descriptive. "pedantic" can be anything, not necessarily formatting-related.

**Comment by jcommelin** [comment]
Otherwise, LGTM

**Comment by adomani** [comment]
Johan, thank you for the comment: I like the new name!

As for the large import, I think that this is to be expected by a linter: it is imported in `Mathlib.Tactic.Linter` and in turn this file is imported virtually everywhere.

**Comment by grunweg** [comment]
Is there anything preventing this linter from landing? All outstanding comments seem to be addressed, CI passes, the linter is disabled by default anyway...

**Comment by adomani** [comment]
I think that all that is missing is a maintainer merging it.  😊

**Comment by grunweg** [comment]
Will this help then? :-)
maintainer merge

**Review by jcommelin — COMMENTED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**
